lib/rhea-promise/index.ts

```diff
@@ -0,0 +1,662 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+import * as rhea from "rhea";
+import * as debugModule from "debug";
+
+const debug = debugModule("rhea-promise");
+
+/**
+ * Eastablishes an amqp connection.
+ * @param {ConnectionOptions} [options] Options to be provided for establishing an amqp connection.
+ * @return {Promise<Connection>} Promise<Connection>
+ * - **Resolves** the promise with the Connection object when rhea emits the "connection_open" event.
+ * - **Rejects** the promise with an AmqpError when rhea emits the "connection_close" event while trying
+ * to establish an amqp connection.
+ */
+export async function connect(options?: ConnectionOptions): Promise<any> {
+  return new Promise((resolve, reject) => {
+    const connection = rhea.connect(options);
+
+    function onOpen(context: any): void {
+      connection.removeListener("connection_open", onOpen);
+      connection.removeListener("connection_close", onClose);
+      connection.removeListener("disconnected", onClose);
+      process.nextTick(() => {
+        debug("Resolving the promise with amqp connection.");
+        resolve(connection);
+      });
+    }
+
+    function onClose(context: Context): void {
+      connection.removeListener("connection_open", onOpen);
+      connection.removeListener("connection_close", onClose);
+      connection.removeListener("disconnected", onClose);
+      debug(`Error occurred while establishing amqp connection.`, context.connection.error);
+      reject(context.connection.error);
+    }
+
+    connection.once("connection_open", onOpen);
+    connection.once("connection_close", onClose);
+    connection.once("disconnected", onClose);
+  });
+}
+
+/**
+ * Closes the amqp connection.
+ * @param {Connection} connection The amqp connection that needs to be closed.
+ * @return {Promise<void>} Promise<void>
+ * - **Resolves** the promise when rhea emits the "connection_close" event.
+ * - **Rejects** the promise with an AmqpError when rhea emits the "connection_error" event while trying
+ * to close an amqp connection.
+ */
+export async function closeConnection(connection: any): Promise<void> {
+  return new Promise<void>((resolve, reject) => {
+    if (!connection || (connection && typeof connection !== "object")) {
+      reject(new Error("connection is a required parameter and must be of tyepe 'object'."));
+    }
+
+    if (connection.is_open()) {
+      function onClose(context: Context): void {
+        connection.removeListener("connection_close", onClose);
+        process.nextTick(() => {
```

Comment:
why the `nextTick`? and if here, why not everywhere else too?

Reply:
remember when things were happening too fast while creating the receiver (i.e.) the server sending attach first and moments later a detach. Hence I added process.nextTick() for all the close and open methods to slow this a little. This may help to catch errors, just in case.